### ISSUE TITLE
An issue with Diagonal pathfinding

### ISSUE THREAD

**Issue by Bru7aLMike** [madyanov/l2mapconv-public#10]
Everything works perfectly. This is like the best geodata creator out there. However, there are some locations in which diagonal pathfinding just hits a brick wall and starts moving either North-South or East-West between the different geo points. Resulting in choppy left-to-right (back and fort) character motion, as opposed to a smooth transition between the two points. Also, the location is quite flat, and it has no blocked geo points, meaning it should be working just fine. And it is not the pathfinding itself, since when I change the geodata file with a High Five one, it does not do that on this specific location (has issues on other places) but this exact no-diagonal movement is gone. Any ideas what could be the culprit of the issue?

SUGGESTION:

And as a follow up on this, a rather flat terrain could use single geodata blocks (like big ones for optimisation purposes) and they should be surrounded by smaller blocks where needed, as shown at the end of the following video. Also, additional options (a config file?) for changing parameters before building a bunch of maps at once would be great!

https://www.youtube.com/watch?v=WdcXrzkkQK0&ab_channel=MapsCrafter

**Comment by madyanov** [comment]
> This is like the best geodata creator out there.

It's not, at least because of #8 :)

>  However, there are some locations in which diagonal pathfinding just hits a brick wall and starts moving either North-South or East-West between the different geo points.

Can you show screenshots of that locations in the _preview_ mode? And may be difference with the working geodata.

> Any ideas what could be the culprit of the issue?

I think it may be problem of the geo engine. It can't handle too many layers or something like this.
On some geo engines all working fine, on some not.

**Comment by Bru7aLMike** [comment]
I will make screenshots, but the other one that I tested (and worked for the most part) had these big blocks of geodata for flat surfaces. They will help reduce the strain on the geo-engine, in my opinion. The way that comes to mind is to make it so that it combines 4by4 adjacent geodata blocks on the same plain as one big block with only 4 directions, as opposed to 64 total directions. And if it is not that hard to do, maybe you can also add a condition, if four blocks (2by2) are adjacent to each other and have Z value in a range lower than the manually set Cell Height, they can also be combined into a single medium-sized block, to further optimise it. The thing is, they should only be in formations 2by2 or 4by4 or 8by8. I'm not a programmer, and I don't even know which language the programme is written on, but at least in Java it is not that hard to several conditions for block-creation (or rather merging). 

DISCLAIMER:

I will edit this post in  3-4 hours with screenshots and a video of both geodata files and the problematic zones.

**Comment by madyanov** [comment]
> had these big blocks of geodata for flat surfaces

l2mapconv also build with this big blocks. They are just not visible in preview mode when building (#6 RU).
You can put .l2j file from `output` directory to `geodata` and check `Imported Geodata` checkmark to see it.

>  And if it is not that hard to do, maybe you can also add a condition

They already merged by `Cell Height`.

> The only half-decent workaround that I've found was to make the geocells big

`Cell Size` is multiplied by 2 (8 * 2 = 16), and 16 is the only accepted cell size for geodata.
The algorithm is not perfect, I don't recommend change `Cell Size`.

**Comment by Bru7aLMike** [comment]
I believe the geodata file from these screenshots was generated in bulk with the rest of the maps, so it uses the default settings, but I could be wrong:
spot 1:
![image](https://user-images.githubusercontent.com/72753019/100750733-21ab3200-33ef-11eb-9d79-1ad833cbcde0.png)
spot 2:
![image](https://user-images.githubusercontent.com/72753019/100750909-5919de80-33ef-11eb-875a-749161255579.png)

And this is how the other, non-problematic geodata looks like:
![image](https://user-images.githubusercontent.com/72753019/100752345-2a9d0300-33f1-11eb-8967-c922564c5a42.png)
(+static mesh to make it easier to recognise what is what)
![image](https://user-images.githubusercontent.com/72753019/100752870-c3338300-33f1-11eb-97b1-c2fe7e100140.png)

![image](https://user-images.githubusercontent.com/72753019/100752611-7b146080-33f1-11eb-9ae8-e2737c7f6956.png)
+staticmesh:
![image](https://user-images.githubusercontent.com/72753019/100752940-db0b0700-33f1-11eb-96e3-d23a14cdc347.png)


!disclaimer - the geodata is from a High Five build, and I've only used it as a reference point. To this end, I've hidden the textures and static meshes for better geodata visibility, since certain parts of the geodata were hidden under the map or a static mesh.
!disclaimer 2 - By non-problematic I mean no zig-zagging and nothing else!



These two areas in particular are causing me headaches. I've determined that the geo engine I'm using, and more specifically the pathfinding buffer, could be one of the causes for the zig-zagging. I'm going over the code now to try and spot where the issue might lie. I have a VIDEO HERE, which shows the zig-zagging issue of the problematic location I referenced as spot 2, However, the geo-engine might not be solely related to the zig-zagging. It might also have something to do with issue #8 in combination with some offset in the geo engine that I'm using. I'm trying to crack the code, since the tool is super convenient and easy to use, and would save me hours upon hours of painful manual labour.

**Comment by madyanov** [comment]
> static meshes for better geodata visibility, since certain parts of the geodata were hidden under the map or a static mesh

That's another problem. Are you sure you use right geodata for client? I've checked many map squares with reference HF geodata (don't know where I did download it tbh), and didn't notice this. Could you provide name of the map square?

> And this is how the other, non-problematic geodata looks like:

There may be difference in the reference grid and generated grid, but I think it's not crucial, since the logic of geoengine should be simple: if direction is passable (not black), player can go from this cell to neighbour. Even if there is no big blocks in resulting geodata, good geoengine should work fine.

**BUT** you can adjust this value https://github.com/madyanov/l2mapconv-public/blob/master/geodata/src/Optimizer.cpp#L11 (SIMPLE_BLOCK_MAX_HEIGHT_DIFFERENCE) to 64 for example, or any other value, to generate more big blocks, and check how it works with your geoengine.

**Comment by Bru7aLMike** [comment]
Thanks.

 And is there a way to make tell the programme to create new cells layers only when the height of a static mesh (or terrain) exceeds (goes out of bounds of) the height of the previous (lower) cell layer?

I've just been to Imperial Tomb and the "carpets" in the first room are very very short in terms of height, and yet, they use a whole new layer. 
![image](https://user-images.githubusercontent.com/72753019/100864961-655d7480-349f-11eb-9a36-b4c02716e8aa.png)

**Comment by madyanov** [comment]
>And is there a way, or us/me the common user, to tell the programme to create new cell layers only when the height of a static mesh (or terrain) exceeds

You can set `Cell Height` to 8 to achieve this.
But it will not work for all cases because of cell height rounding.
Original HF geodata also may have such issues, but not so often.

**Comment by Bru7aLMike** [comment]
The cell height on that particular image is either 8 or 16, along with the cell size. I keep them identical since it helps reduce the zig-zagging, albeit only by a small margin. This ultimately means that there might be an error in the height calculations of the geodata generator.
 I've tested several different geodatas on the same server (same geoengine). Two were geodata files for High Five chronicle (so they are not accurate, in terms of heightmap and static mesh. However, I was able to identify the reasons for the zig-zagging and when it occurs. So, I found the same patterns that were causing the zig-zagging and started testing. Neither of the other geodatas produced any zig-zagging. I also tested a geodata file made for the same chronicle, but different iteration, so there were misplaced static meshes, just as with the High Five geodata - no problems either.  I also tested a manually created geodata version for the specific map, (it is a lengthy process and involves a lot of conversions, so it is laborious and inconvenient, to say the least, and full geodata production would take more than 50 man-hours). And even the manually created (meaning non-professionally made) map did not produce any zig-zagging. I'm still trying to find a way to fix the zig-zagging through the geoengine and pathfinding javas, inspecting them for potential code/math flaws, but the fact that all other geodatas had no zig-zag problems bring me back to you. Any ideas what might be the cause of my issue? And it only happens when traversing diagonally. NSWE directions are just fine.
